### DISCUSSION TITLE
Why is there an `.envrc` in the repository?

### DISCUSSION THREAD

**Discussion by oderwat** [helix-editor/helix#12313] [General]
I just was asked to allow the `.envrc` from this repository. I wonder why there is a `.envrc` and what about all those Nix related stuff? Do I need nix to compile the program? Do I require it for making contributions? I am new to a Rust codebase and really wonder about the level of added dependencies in the environment setup.

**Comment by TornaxO7** [comment]
> I just was asked to allow the .envrc from this repository. I wonder why there is a .envrc and what about all those Nix related stuff?

`.envrc` is coming from [`direnv`](https://direnv.net/). Basically, in this case, it will automatically execute `nix develop` if you are entering the directory. This is pretty nice for users who are using the nix package manager because you just simply need to allow it, enter in and you have all tools installed to immediately start working with the project.

> Do I need nix to compile the program? Do I require it for making contributions?

![peepoNo](https://github.com/user-attachments/assets/c52a3ddf-494a-425c-9566-2c4dea5e7a96)

It (nix) is just a convenient setup if you are familiar with the tool. Nothing else.

**Comment by zetashift** [comment]
It's a file that allows for the `direnv` (https://direnv.net/) to automatically load environment variables/set stuff in the shell when you enter the directory that contains it. These things then get unloaded when you leave the directory or its subdirs, e.g. `cd ..`.

You do not need nix to contribute to the codebase, it's opt-in, if you disallow the loading of the `.envrc` or you never install Nix, nothing will happen. You can install rust via `rustup` and contribute)

The Nix stuff sets up the development dependencies and allows one to get started with `cargo build` without looking on the internet for the right dependencies(and versions of said deps).

> really wonder about the level of added dependencies in the environment setup.

I understand this perspective, but as @TornaxO7 said, it's just for convenience.

(btw this is probably better off as a discussion)

**Comment by oderwat** [comment]
I know where it comes from (obviously) and wonder why it is forced onto contributors. I do not use Nix and would probably add my own `.envrc`.

**Comment by TornaxO7** [comment]
> I know where it comes from (obviously) and wonder why it is forced onto contributors. 

It's not forced, you can simply disallow it or am I missunderstanding you?

**Comment by oderwat** [comment]
When I disallow it, I still can't use my own one (and get warnings all the time). I use asdf and Nushell btw.  and have a different approach in managing my dependencies.

**Reply by kirawi** [reply]
~That's surprising, what tooling causes this?~

Nevermind, I understand now.

**Comment by oderwat** [comment]
As maintainer, I would probably add a `.envrc-nix` and let users that want to use it add a symlink as `.envrc`.

**Comment by zetashift** [comment]
> When I disallow it, I still can't use my own one (and get warnings all the time). I use asdf and Nushell btw. and have a different approach in managing my dependencies.

Ah I see, I misunderstood the original post then. 
I'll wait 'till the official maintainers share their perspective, personally I'd prefer it if `.envrc` was not commited as well, and I say that while using Nix!

**Comment by pascalkuthe** [comment]
You can edit the file there is really no problem here

**Reply by oderwat** [reply]
> You can edit the file there is really no problem here

There is no problem for you, but for somebody that does not need or want to change the file it will create problems as it will show this files as changed. You have to watch out not to add it to commits, and it will irritate tooling. Enforcing your environment to others and telling them, that there is no issue feels rude to me. This is why I asked if it is mandatory to use Nix for contributing. It probably was too subtle of a hint.

**Reply by SrEstegosaurio** [reply]
Nobody is enforcing any environment on anyone, you can set it up however you prefer locally. At the end of the day we're talking about a one line change. And if we were to be really pedantic your argument could be easily reversed to argue against your point.

Cool part being you can do whatever you want locally. :)

**Reply by oderwat** [reply]
> about a one line change

What are you talking about? Why is changing a file a "one line change" in your eyes?

Also: Please use my argument against me! I wonder how that goes.

So much ignorance from seasoned developers irritates me. Of course can I change that file. But it is part of the git repository and will therefor always show up as modified and needs to be dealt with on commits and possibly when pulling.

What is the big problem in following some sane conventions on respecting other developers systems?

It is so easy: Rename that file to a template (`.envrc-nix-flake`), add  `.envrc` to `.gitignore` and let people that want to use this `.envrc` just create a copy or symlink of that template file?

**Reply by TornaxO7** [reply]
> It is so easy: Rename that file to a template (.envrc-nix-flake), add .envrc to .gitignore and let people that want to use this .envrc just create a copy or symlink of that template file?

Well that sounds like a good idea.
If the maintainers agree, I can create a PR for it.
